### ISSUE TITLE
Assertion error triggered by CrossRefFakeMode

### ISSUE THREAD

**Issue by jgong5** [pytorch/pytorch#90652]
### 🐛 Describe the bug

Running with `TORCH_COMPILE_DEBUG=1`, if the model contains ops without meta tensor support, I got the failure due to assertion from
https://github.com/pytorch/pytorch/blob/7ec1cb855373dbc011b403adb87d943e395d8d2a/torch/_subclasses/meta_utils.py#L203-L205

The issue can be demonstrated with a simple test case below:
```python
import torch
from torch.utils._mode_utils import no_dispatch

with torch._subclasses.FakeTensorMode():
    a = torch.rand([100])
    with torch._subclasses.CrossRefFakeMode():
        # below logic from run_fallback_kernel
        with no_dispatch():
            b = torch.zeros_like(a) # <- exception thrown here
```
The chain looks something like `FakeTensorMode` -> `run_fallback_kernel` -> `no_dispatch()` -> `torch.zeros_like` -> `CrossRefFakeMode` -> `from_tensor` -> ... -> assertion error expecting `no_dispatch` off.

The error is triggered from `CrossRefFakeMode` which will be turned on if the `TORCH_COMPILE_DEBUG=1` is on.

My question:
1. What's the contract of calling functions on tensor subclasses with __torch_dispatch__ with `no_dispatch`? Is it allowed to call `torch.zeros_like(a)` under `no_dispatch` when `a` is a tensor subclass like proxy tensor?
2. What's the purpose of `CrossRefFakseMode`? Is it expecting real tensors or can also support fake tensors?

### Versions

Collecting environment information...
PyTorch version: 2.0.0a0+git184f6b5
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.11.0-27-generic-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] clip-anytorch==2.5.0
[pip3] CoCa-pytorch==0.0.6
[pip3] dalle2-pytorch==1.10.5
[pip3] ema-pytorch==0.0.10
[pip3] functorch==1.14.0a0+408bcf1
[pip3] intel-extension-for-pytorch==1.13.0+cpu
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] pytorch-transformers==1.2.0
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.1.5
[pip3] torch==2.0.0a0+git7552e12
[pip3] torch-fidelity==0.3.0
[pip3] torch-struct==0.5
[pip3] torchaudio==0.14.0a0+41c4fd6
[pip3] torchmetrics==0.10.3
[pip3] torchrec-nightly==2022.11.21
[pip3] torchtext==0.15.0a0+58dad70
[pip3] torchvision==0.15.0a0+3100734
[pip3] torchx==0.3.0
[pip3] vector-quantize-pytorch==0.10.12
[conda] bert-pytorch              0.0.1a4                   dev_0    <develop>
[conda] clip-anytorch             2.5.0                    pypi_0    pypi
[conda] coca-pytorch              0.0.6                    pypi_0    pypi
[conda] dalle2-pytorch            1.10.5                   pypi_0    pypi
[conda] ema-pytorch               0.0.10                   pypi_0    pypi
[conda] functorch                 1.14.0a0+408bcf1          pypi_0    pypi
[conda] intel-extension-for-pytorch 1.13.0+cpu                dev_0    <develop>
[conda] mkl                       2022.1.0           hc2b9512_224  
[conda] mkl-include               2022.2.1                 pypi_0    pypi
[conda] mkl-static                2022.2.1                 pypi_0    pypi
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch-transformers      1.2.0                    pypi_0    pypi
[conda] pytorch-warmup            0.1.1                    pypi_0    pypi
[conda] rotary-embedding-torch    0.1.5                    pypi_0    pypi
[conda] torch                     2.0.0a0+git7552e12           dev_0    <develop>
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torch-struct              0.5                      pypi_0    pypi
[conda] torchaudio                0.14.0a0+41c4fd6          pypi_0    pypi
[conda] torchmetrics              0.10.3                   pypi_0    pypi
[conda] torchrec-nightly          2022.11.21               pypi_0    pypi
[conda] torchtext                 0.15.0a0+58dad70          pypi_0    pypi
[conda] torchvision               0.15.0a0+3100734          pypi_0    pypi
[conda] torchx                    0.3.0                    pypi_0    pypi
[conda] vector-quantize-pytorch   0.10.12                  pypi_0    pypi

**Comment by ezyang** [comment]
CrossRefFakeMode shouldn't be run on the inside of FakeMode; the way it works is you run a normal model, we run the fake compute along side it and check if fake is implemented correctly.

So maybe this is a misimplementation in TORCH_COMPILE_DEBUG

**Comment by jgong5** [comment]
@ezyang Is it expected that `__torch_dispatch__` is invoked on some tensor subclasses under `no_dispatch` context? Like the code in `run_fallback_kernel`:

https://github.com/pytorch/pytorch/blob/8fd31ac4dade93a10822ba32c1f2293827b2b2f3/torch/_subclasses/fake_tensor.py#L1050

**Comment by ezyang** [comment]
I keep forgetting what no_dispatch does; @samdow would know. It's possible it turns off tensor dispatch but not mode dispatch.

**Comment by jgong5** [comment]
> no_dispatch adds the exclude guard for the python key (meaning it doesn't run torch dispatch for subclass nor modes) so I'm really curious what's happening here. Especially since without the CrossRefFakeMode, I'm seeing that `b` is a vanilla tensor (maybe good, maybe bad)

`e` is a fake tensor and I saw `__torch_dispatch__` was invoked on the mode on top of the current FakeTensorMode, e.g., CrossRefFakeMode. This seems to break the statement: "meaning it doesn't run torch dispatch for subclass nor modes".
